text/0000-min-rust-version.md

```diff
@@ -0,0 +1,179 @@
+
+`cargo publish` will check that upload is done with a toolchain version specified
+in the `rust` field. If toolchain version is different, `cargo` will refuse to
+upload the crate. It will be a failsafe to prevent uses of incorrect `rust` values
+due to unintended MSRV bumps. This check can be disabled by using the existing
+`--no-verify` option.
+
+## Making `rust` field mandatory
+
+In future (probably in a next edition) we could make `rust` field mandatory for
+a newly uploaded crates. MSRV for older crates will be determined by the `edition`
+field. In other words `edition = "2018"` will imply `rust = "1.31"` and
+`edition = "2015"` will imply `rust = "1.0"`.
+
+`cargo init` will use version of a currently used toolchain.
+
+## `cfg`-based MSRVs
+
+Some crates can have different MSRVs depending on target architecture or enabled
+features. In such cases it can be useful to describe how MSRV depends on them,
+e.g. in the following way:
+```toml
+[package]
+rust = "1.30"
+
+[target.x86_64-pc-windows-gnu]
+rust = "1.35"
+
+[target.'cfg(feature = "foo")']
+rust = "1.33"
+```
+
+All `rust` values in the `target` sections should be equal or bigger to a `rust` value
+specified in the `package` section.
+
+If target condition is true, then `cargo ` will use `rust` value from this section.
+If several target section conditions are true, then maximum value will be used.
+
+## Nightly and stable versions
+
+Some crates may prefer to target only the most recent stable or nightly toolchain.
+In addition to the versions we could allow `stable` and `nightly` values to declare
+that maintainers do not track MSRV for the crate.
+
+For some bleeding-edge crates which experience frequent breaks on Nightly updates
+(e.g. `rocket`) it can be useful to specify exact Nightly version(s) on which
+crate can be built. One way to achieve this is by using the following syntax:
+- auto-select: "nightly" This variant will behave in the same way as "stable", i.e.
+it will take a current nightly version and will use it in a "more or equal" constraint.
+- single version: "nightly: 2018-01-01" (the main variant)
+- enumeration: "nightly: 2018-01-01, 2018-01-15"
+- semver-like conditions: "nightly: >=2018-01-01", "nightly: >=2018-01-01, <=2018-01-15",
+"nightly: >=2018-01-01, <=2018-01-15, 2018-01-20". (the latter is interpreted as
+"(version >= 2018-01-01 && version <= 2018-01-20) || version == 2018-01-20")
+
+Such restrictions can be quite severe, but hopefully this functionality will be
+used only by handful of crates.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Declaration of MSRV, even with the checks, does not guarantee that crate
+will work correctly on the specified MSRV, only appropriate CI testing can do that.
+- More complex dependency versions resolution algorithm.
+- MSRV selected by `cargo publish` with `rust = "stable"` can be too
+conservative.
+
+# Alternatives
+[alternatives]: #alternatives
+
+- Automatically calculate MSRV.
```

Comment:
This seems too vague to have much meaning; there are wildly different things it could mean, with completely different trade-offs. What is automatically calculating the MSRV? rustc, cargo, crates.io? How is it conveyed? Would it play well with the future work of MSRV influencing version resolution? I can think of two radically different approaches to this off the top of my head, one of which is completely different from this proposal and one of which boils down to just augmenting the `package.rust` field of this specification with some tooling to automate discovery of the MSRV.

Reply:
I didn't want to cover all potential approaches, as you do say yourself there are a lot of options to choose from. The main idea here is that instead of asking people to manually select MSRV via the `rust` field, we could rely on some automatic system (be it on rustc, cargo or crates.io side) to partially solve issues which have motivated this RFC.